platforms/core-runtime/build-configuration/src/main/java/org/gradle/internal/buildconfiguration/resolvers/DefaultToolchainRepositoriesResolver.java

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+
+package org.gradle.internal.buildconfiguration.resolvers;
+
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.JavaToolchainDownload;
+import org.gradle.jvm.toolchain.JavaToolchainRequest;
+import org.gradle.jvm.toolchain.JavaToolchainResolverRegistry;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
+import org.gradle.jvm.toolchain.JvmImplementation;
+import org.gradle.jvm.toolchain.JvmVendorSpec;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainRequest;
+import org.gradle.jvm.toolchain.internal.DefaultToolchainSpec;
+import org.gradle.jvm.toolchain.internal.JavaToolchainResolverRegistryInternal;
+import org.gradle.jvm.toolchain.internal.RealizedJavaToolchainRepository;
+import org.gradle.platform.BuildPlatform;
+
+import javax.annotation.Nullable;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.gradle.internal.buildconfiguration.resolvers.ToolchainSupportedPlatformsMatrix.getToolchainSupportedBuildPlatforms;
+
+public class DefaultToolchainRepositoriesResolver implements ToolchainRepositoriesResolver {
+
+    private final JavaToolchainResolverRegistryInternal toolchainResolverRegistry;
+    private final ObjectFactory objectFactory;
+
+    public DefaultToolchainRepositoriesResolver(JavaToolchainResolverRegistry toolchainResolverRegistry, ObjectFactory objectFactory) {
+        this.toolchainResolverRegistry = (JavaToolchainResolverRegistryInternal) toolchainResolverRegistry;
+        this.objectFactory = objectFactory;
+    }
+
+    @Override
+    public Map<BuildPlatform, Optional<URI>> resolveToolchainDownloadUrlsByPlatform(
+        JavaLanguageVersion toolchainVersion,
+        @Nullable String toolchainVendor,
+        @Nullable JvmImplementation toolchainImplementation
+    ) throws UnconfiguredToolchainRepositoriesResolver {
+        List<? extends RealizedJavaToolchainRepository> toolchainRepositories = toolchainResolverRegistry.requestedRepositories();
+        if (toolchainRepositories.isEmpty()) {
+            throw new UnconfiguredToolchainRepositoriesResolver();
+        }
+
+        JavaToolchainSpec toolchainSpec = createToolchainSpec(toolchainVersion, toolchainVendor, toolchainImplementation);
+        Map<BuildPlatform, Optional<URI>> toolchainDownloadUrlByPlatformMap = new HashMap<>();
+
+        getToolchainSupportedBuildPlatforms().stream()
+            .map(buildPlatform -> new DefaultJavaToolchainRequest(toolchainSpec, buildPlatform))
+            .forEach(javaToolchainRequest -> {
+                Optional<URI> downloadUrl = resolveToolchainDownloadUrlRequest(toolchainRepositories, javaToolchainRequest);
+                toolchainDownloadUrlByPlatformMap.put(javaToolchainRequest.getBuildPlatform(), downloadUrl);
+            });
+
+        return toolchainDownloadUrlByPlatformMap;
```

Comment:
Will be great in the case the applied project resolver plugins aren't able to provide any download URLs given a daemon toolchain criteria to throw an exception instead of generating an `gradle/gradle-daemon-jvm.properties`. This is manly to avoid users getting in wrong state were the specified criteria is completely invalid and as a consequence not longer being able to build.